crates/module-system/sov-solana-offchain-auth/src/authentication.rs

```diff
@@ -0,0 +1,483 @@
+            runtime_call: self.runtime_call,
+            uniqueness: self.uniqueness,
+            details: self.details,
+        }
+    }
+
+    fn unmetered_deserialize_inner(buf: &mut &[u8]) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice::<SolanaOffchainUnsignedTransaction<R, S>>(buf)
+    }
+}
+
+impl<R, S> MeteredJsonDeserialize<S> for SolanaOffchainUnsignedTransaction<R, S>
+where
+    S: Spec,
+    R: TransactionCallable,
+    <R as TransactionCallable>::Call: Serialize + DeserializeOwned,
+{
+    fn bias_json_deserialization() -> <S as Spec>::Gas {
+        S::tx_bias_json_deserialization()
+    }
+
+    fn gas_to_charge_per_byte_json_deserialization() -> <S as Spec>::Gas {
+        S::tx_gas_to_charge_per_byte_json_deserialization()
+    }
+
+    #[cfg_attr(feature = "bench", crate::cycle_tracker)]
+    #[cfg_attr(
+        all(feature = "gas-constant-estimation", feature = "native"),
+        crate::track_gas_constants_usage
+    )]
+    fn deserialize(
+        buf: &mut &[u8],
+        meter: &mut impl GasMeter<Spec = S>,
+    ) -> Result<Self, MeteredJsonDeserializeError<<S as GasSpec>::Gas>> {
+        Self::charge_gas_to_deserialize(buf, meter)?;
+
+        SolanaOffchainUnsignedTransaction::<R, S>::unmetered_deserialize_inner(buf)
+            .map_err(MeteredJsonDeserializeError::SerdeError)
+    }
+
+    #[cfg(feature = "native")]
+    fn unmetered_deserialize(
+        buf: &mut &[u8],
+    ) -> Result<Self, MeteredJsonDeserializeError<<S as GasSpec>::Gas>> {
+        SolanaOffchainUnsignedTransaction::<R, S>::unmetered_deserialize_inner(buf)
+            .map_err(MeteredJsonDeserializeError::SerdeError)
+    }
+}
+
+/// The envelope for a signed spec-compliant solana offchain message, where the signed message
+/// includes the preamble.
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct SolanaOffchainSpecCompliantMessage<S: Spec> {
+    /// The message is a JSON-serialized SolanaOffchainUnsignedTransaction with the standard preamble prepended.
+    pub signed_message: Vec<u8>,
+    pub signature: <S::CryptoSpec as CryptoSpec>::Signature,
+}
+
+/// The envelope for a message signed "raw", without the preable included.
+/// The preamble always starts with the \xff byte, whereas our raw message is JSON and so can only
+/// start with an ASCII character (normally, '{'), allowing us to unambiguously differentiate them.
+/// Without the preamble present, we need to include the pubkey explicitly.
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct SolanaOffchainSimpleMessage<S: Spec> {
+    /// The message is a JSON-serialized SolanaOffchainUnsignedTransaction, unaltered.
+    pub signed_message: Vec<u8>,
+    pub pubkey: <S::CryptoSpec as CryptoSpec>::PublicKey,
+    pub signature: <S::CryptoSpec as CryptoSpec>::Signature,
+}
+
+/// The length of a preamble with a single 32-byte signer.
+pub const PREAMBLE_LEN: usize = std::mem::size_of::<RawSolanaOffchainMessagePreamble>();
```

Comment:
I don't think Rustc guarantees that `size_of` will be consistent across platforms or compiler versions, since it depends on the struct packing which is not guaranteed to be stable. This looks like it's working for now, but we may want to revisit if we run into issues.

Reply:
That's a great point, thanks for spotting!
The preamble size is constant, so I'll just hardcode it to 85 then, rather than trying to be too fancy.